apps/deduplication/tasks/indexing.py

```diff
@@ -0,0 +1,148 @@
+import pickle
+from django.db import transaction
+from django.utils import timezone
+from celery import shared_task
+from celery.utils.log import get_task_logger
+from datasketch import LeanMinHash, MinHashLSH
+
+from utils.decorators import warn_on_exception
+from utils.common import batched
+from lead.models import Lead
+from project.models import Project
+from deduplication.models import LSHIndex
+from deduplication.utils import get_minhash, insert_to_index
+
+logger = get_task_logger(__name__)
+
+
+def find_and_set_duplicate_leads(index: MinHashLSH, lead: Lead, minhash: LeanMinHash):
+    duplicate_lead_ids = index.query(minhash)
+    duplicate_leads = Lead.objects.filter(pk__in=duplicate_lead_ids)
+    lead.duplicate_leads.set(duplicate_leads)
+
+
+def process_and_index_lead(lead: Lead, index: MinHashLSH):
+    text = lead.leadpreview.text_extract if hasattr(lead, 'leadpreview') else lead.text
+    if not text:
+        return index
+    minhash = get_minhash(text)
+    find_and_set_duplicate_leads(index, lead, minhash)
+
+    insert_to_index(index, lead.id, minhash)
+    lead.is_indexed = True
+    lead.indexed_at = timezone.now()
+    lead.save(update_fields=['is_indexed', 'indexed_at'])
+    return index
+
+
+def process_and_index_leads(
+    project: Project,
+    index_obj: LSHIndex,
+):
+    # Fetch leads which have been extracted and which have not been indexed
+    leads = Lead.objects.filter(
+        project=project,
+        is_indexed=False,
+    )
+    index: MinHashLSH = index_obj.index
+    try:
+        batches = batched(leads, batch_size=200)
+        for batch in batches:
+            with transaction.atomic():
+                for lead in batch:
+                    process_and_index_lead(lead, index)
+                # Update the index
+                index_obj.index = index
+                index_obj.save()
+    except Exception:
+        logger.warning(
+            f"Error creating index for project {project.title}({project.id})",
+            exc_info=True
+        )
+        import traceback
+
+        index_obj.has_errored = True
+        index_obj.error = traceback.format_exc()
```

Comment:
Maybe use sentry instead of this?

Reply:
I've left it as it is while adding `logger.error` instead of `logger.warning`.

Reply:
Removed.